File: evals/refs/nbody.py
import sys

def accel(p, mass, n, G=1.0):
    a = [[0.0, 0.0] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if i == j: continue
            dx = p[j][0] - p[i][0]
            dy = p[j][1] - p[i][1]
            r2 = dx*dx + dy*dy
            r = r2**0.5
            if r > 1e-12:
                f = G * mass[j] / (r2 * r)
                a[i][0] += f * dx
                a[i][1] += f * dy
    return a

data = sys.stdin.read().split()
idx = 0
T = int(data[idx]); idx += 1
out = []
for _ in range(T):
    N = int(data[idx]); idx += 1
    dt = float(data[idx]); idx += 1
    steps = int(data[idx]); idx += 1
    mass = []
    pos = []
    vel = []
    for _ in range(N):
        m = float(data[idx]); idx += 1
        x = float(data[idx]); idx += 1
        y = float(data[idx]); idx += 1
        vx = float(data[idx]); idx += 1
        vy = float(data[idx]); idx += 1
        mass.append(m); pos.append([x, y]); vel.append([vx, vy])
    a = accel(pos, mass, N)
    for _ in range(steps):
        for i in range(N):
            pos[i][0] += vel[i][0]*dt + 0.5*a[i][0]*dt*dt
            pos[i][1] += vel[i][1]*dt + 0.5*a[i][1]*dt*dt
        a_new = accel(pos, mass, N)
        for i in range(N):
            vel[i][0] += 0.5*(a[i][0] + a_new[i][0])*dt
            vel[i][1] += 0.5*(a[i][1] + a_new[i][1])*dt
        a = a_new
    for i in range(N):
        out.append(f'{pos[i][0]:.6f} {pos[i][1]:.6f} {vel[i][0]:.6f} {vel[i][1]:.6f}')
sys.stdout.write('\n'.join(out) + '\n')

File: evals/refs/nbody.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
)

func accel(p [][]float64, mass []float64, n int) [][]float64 {
	a := make([][]float64, n)
	for i := range a {
		a[i] = []float64{0, 0}
	}
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			if i == j {
				continue
			}
			dx := p[j][0] - p[i][0]
			dy := p[j][1] - p[i][1]
			r2 := dx*dx + dy*dy
			r := math.Sqrt(r2)
			if r > 1e-12 {
				f := mass[j] / (r2 * r)
				a[i][0] += f * dx
				a[i][1] += f * dy
			}
		}
	}
	return a
}

func main() {
	in := bufio.NewScanner(os.Stdin)
	in.Buffer(make([]byte, 1<<20), 1<<20)
	in.Split(bufio.ScanWords)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()
	read := func() string { in.Scan(); return in.Text() }
	T, _ := strconv.Atoi(read())
	for c := 0; c < T; c++ {
		N, _ := strconv.Atoi(read())
		dt, _ := strconv.ParseFloat(read(), 64)
		steps, _ := strconv.Atoi(read())
		mass := make([]float64, N)
		pos := make([][]float64, N)
		vel := make([][]float64, N)
		for i := 0; i < N; i++ {
			m, _ := strconv.ParseFloat(read(), 64)
			x, _ := strconv.ParseFloat(read(), 64)
			y, _ := strconv.ParseFloat(read(), 64)
			vx, _ := strconv.ParseFloat(read(), 64)
			vy, _ := strconv.ParseFloat(read(), 64)
			mass[i] = m; pos[i] = []float64{x, y}; vel[i] = []float64{vx, vy}
		}
		a := accel(pos, mass, N)
		for s := 0; s < steps; s++ {
			for i := 0; i < N; i++ {
				pos[i][0] += vel[i][0]*dt + 0.5*a[i][0]*dt*dt
				pos[i][1] += vel[i][1]*dt + 0.5*a[i][1]*dt*dt
			}
			aNew := accel(pos, mass, N)
			for i := 0; i < N; i++ {
				vel[i][0] += 0.5 * (a[i][0] + aNew[i][0]) * dt
				vel[i][1] += 0.5 * (a[i][1] + aNew[i][1]) * dt
			}
			a = aNew
		}
		for i := 0; i < N; i++ {
			fmt.Fprintf(out, "%.6f %.6f %.6f %.6f\n", pos[i][0], pos[i][1], vel[i][0], vel[i][1])
		}
	}
}
